docs: Documented Send and Sync requirements for Mutex + MutexGuard

ranger-ross opened the pull request:
<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

This an attempt to continue where #123225 left off.

I did some light clean up from the work done in that PR.
I also documented the `!Send` + `Sync` implementations for `MutexGuard` to the best of my knowledge.
Let me know if I got anything wrong :smile: 

fixes #122856

cc: @IoaNNUwU

r? @joboet

ranger-ross committed:
docs: Documented Send and Sync requirements for Mutex + MutexGuard

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/12844879970/job/35818356140) [(plain)](https://github.com/rust-lang/rust/commit/068039bd8ad466d9827d71d170031ddc80fc1d1a/checks/35818356140/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 Documenting std v0.0.0 (/checkout/library/std)
error: unresolved link to `Rc`
   --> library/std/src/sync/poison/mutex.rs:191:43
    |
191 | /// this manner. For instance, consider [`Rc`], a non-atomic reference counted smart pointer,
    |                                           ^^ no item named `Rc` in scope
    = help: to escape `[` and `]` characters, add '\' before them like `\[` or `\]`
    = note: `-D rustdoc::broken-intra-doc-links` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(rustdoc::broken_intra_doc_links)]`


error: could not document `std`
warning: build failed, waiting for other jobs to finish...
Command has failed. Rerun with -v to see more details.
  local time: Sat Jan 18 14:59:05 UTC 2025
  network time: Sat, 18 Jan 2025 14:59:05 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

joboet commented on library/std/src/sync/poison/mutex.rs:
```diff
@@ -181,10 +181,26 @@ pub struct Mutex<T: ?Sized> {
-// these are the only places where `T: Send` matters; all other
-// functionality works fine on a single thread.
+/// Mutex is a container that wraps `T`, so it's necessary for `T` to be `Send`
+/// to safely send `Mutex` to another thread. This ensures that the protected
+/// data can be accessed safely from multiple threads without causing data races
+/// or other unsafe behavior.
```
The part about multiple threads is relevant for `Sync`. If you rely on `Send`, you have exclusive access anyways.

joboet commented on library/std/src/sync/poison/mutex.rs:
```diff
@@ -181,10 +181,26 @@ pub struct Mutex<T: ?Sized> {
+/// data can be accessed safely from multiple threads without causing data races
+/// or other unsafe behavior.
+///
+/// [`Mutex<T>`] provides mutable access to `T` to one thread at a time. However, it's essential
+/// for `T` to be `Send` because it's not safe for non-`Send` structures to be accessed in
+/// this manner. For instance, consider [`Rc`], a non-atomic reference counted smart pointer,
+/// which is not `Send`. With `Rc`, we can have multiple copies pointing to the same heap
+/// allocation with a non-atomic reference count. If we were to use `Mutex<Rc<_>>`, it would
+/// only protect one instance of `Rc` from shared access, leaving other copies vulnerable
+/// to potential data races.
+///
+/// [`Rc`]: crate::rc::Rc
```
The same is applicable here, this is relevant for `Sync` but not for `Send`. Perhaps you could phrase this around `Mutex::new` and `Mutex::into_inner` instead?

joboet commented on library/std/src/sync/poison/mutex.rs:
```diff
@@ -211,8 +227,18 @@ pub struct MutexGuard<'a, T: ?Sized + 'a> {
 impl<T: ?Sized> !Send for MutexGuard<'_, T> {}
+
+/// A [`MutexGuard`] can be `Sync` even though it is not `Send` because ownership is not transfer to
+/// a new thread. Because ownership always stays on the original thread `Sync`
+/// is safe to implement for [`MutexGuard`].
+/// See the `!Send` implementation on [`MutexGuard`] for more details.
```
I don't think ownership is very relevant here. The important part is that you can get a `&T` from `&MutexGuard<T>`, so as long as `T` is `Sync` (and therefore `&T: Send`), `MutexGuard<T>` can also be `Sync`.

ranger-ross replied:
Thanks, I think this is a much better way of framing it.

joboet reviewed (changes requested):
Thank you for picking this up, it would be great to have some documentation here.

ranger-ross commented:
@rustbot label -S-waiting-on-author +S-waiting-on-review

joboet reviewed (commented):
Sorry for the language-lawyering. I'm studying a humanities subject, so I can be a bit picky about this stuff... 😄

joboet commented:
@rustbot author

ranger-ross commented:
> Sorry for the language-lawyering. I'm studying a humanities subject, so I can be a bit picky about this stuff... 😄

No worries. I was expecting a fairly stringent for this. 😆
Thanks for the feedback.

I believe I addressed your comments above so let me know what you think. :)

@rustbot reviewer

joboet commented on library/std/src/sync/poison/mutex.rs:
```diff
@@ -181,10 +181,29 @@ pub struct Mutex<T: ?Sized> {
+/// only protect one instance of `Rc` from shared access, leaving other copies vulnerable
+/// to potential data races.
+///
+/// Also note that it is not necessary for `T` to be `Sync` as it is not possible
+/// to acquire a `&T` from a [`Mutex`].
```
It is possible to get a `&T` by dereferencing a mutex guard, so this is misleading. The argument here is that the `&T` is only made available to one thread at a time (if `T` is not `Sync`). This is essentially equivalent to sending a `&mut T` between threads and creating a `&T` from that.

ranger-ross replied:
Ahh, my bad. I updated this doc based on your comment above. (it was basically a copy/paste since I think your explanation is quiet good)

joboet commented on library/std/src/sync/poison/mutex.rs:
```diff
@@ -211,8 +230,16 @@ pub struct MutexGuard<'a, T: ?Sized + 'a> {
     poison: poison::Guard,
 }
 
+/// A [`MutexGuard`] is not `Send` to maximize platform portablity.
```
```suggestion
/// A [`MutexGuard`] is not `Send` to maximize platform portablity.
///
```

joboet reviewed (changes requested):
This reads very nicely. I still have one concern and one nit, but otherwise, this seems ready to merge.

ranger-ross commented:
@rustbot reviewer

joboet commented:
Thank you for the PR!
@bors r+ rollup

bors commented:
:pushpin: Commit 3d84a49c376067c67d878b6710adc9865d758c62 has been approved by `joboet`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=joboet 3d84a49c376067c67d878b6710adc9865d758c62 -->
<!-- homu: {"type":"Approved","sha":"3d84a49c376067c67d878b6710adc9865d758c62","approver":"joboet","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
